ParticleObjects/Source/mixSource_class.f90

```diff
@@ -0,0 +1,160 @@
+module mixSource_class
+
+  use numPrecision
+  use errors_mod,        only : fatalError
+  use genericProcedures, only : numToChar, linearSearchFloor
+  use dictionary_class,  only : dictionary
+  use RNG_class,         only : RNG
+  use particle_class,    only : particleState
+
+  ! Source
+  use source_inter,       only : source
+
+  ! Geometry
+  use geometry_inter,    only : geometry
+
+  implicit none
+  private
+
+  !!
+  !! Public Pointer Type Cast
+  !! Needed for initialisation in sourceFactory_func
+  !!
+  public mixSource_TptrCast
+
+  !!
+  !! Helper type to store polymorphic instances of sources
+  !!
+  type source_slot
+    class(source), allocatable :: sourceItem
+  end type source_slot
+
+  !!
+  !! Class describing a mixture of sources
+  !!
+  !! It builds and samples from a discrete cdf which source the particle should be
+  !! sampled from. The cdf is built from the sources' weights, provided by the user.
+  !!
+  !! Private members:
+  !!   sources -> array of sources
+  !!   cdf     -> cdf used to select a source
+  !!
+  !! Interface:
+  !!   init              -> initialise mixture of sources
+  !!   sampleParticle    -> sample which source to sample the particle from
+  !!   kill              -> terminate source
+  !!
+  !! Sample Dictionary Input:
+  !!   source {
+  !!       type mixSource;
+  !!       sources (src1 src2);
+  !!       weights (0.5 0.5);
+  !!       src1 { < source 1 definition> }
+  !!       src1 { < source 2 definition> }
+  !!      }
+  !!
+  type, public, extends(source) :: mixSource
+    type(source_slot), dimension(:), allocatable :: sources
+    real(defReal), dimension(:), allocatable     :: cdf
+  contains
+    procedure :: init
+    procedure :: sampleParticle
+    procedure :: kill
+  end type mixSource
+
+contains
+
+  !!
+  !! Initialise from dictionary
+  !!
+  !! See source_inter for details
+  !!
+  !! Errors:
+  !!   - when the number of weigths doesn't correspond to the number of sources
+  !!   - if any of the weights is zero
+  !!
+  subroutine init(self, dict, geom)
+    class(mixSource), intent(inout)          :: self
+    class(dictionary), intent(in)            :: dict
+    class(geometry), pointer, intent(in)     :: geom
+    integer(shortInt)                        :: i, N
+    real(defReal), dimension(:), allocatable :: weights
+    character(100), parameter :: Here = 'init (mixSource_class.f90)'
+
+    ! Get number of sources
+    N = size(self % sources)
```

Comment:
This is a bit bewildering. Might be good to highlight that sources has already been initialised elsewhere due to trying to avoid circular dependencies.